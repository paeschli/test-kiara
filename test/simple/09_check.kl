# Function to print a double.
extern [C] printd(x:double):double;

def intrinsic [llvm] printdd(a:double):double
"@.str = private unnamed_addr constant [12 x i8] c\"value = %f\\0A\\00\", align 1 "
"declare i32 @printf(i8* nocapture, ...) nounwind "

"define double @${mangledName}(double %a) nounwind uwtable { "
"entry: "
"  %call = tail call i32 (i8*, ...)* @printf(i8* getelementptr inbounds ([12 x i8]* @.str, i64 0, i64 0), double %a) nounwind "
"  ret double 0.000000e+00 "
"}"
;

def intrinsic [llvm] addii(a:int, b:int):int
    "define i32 @${mangledName}(i32 %a, i32 %b) nounwind uwtable readnone {"
    "entry:"
    "  %add = add nsw i32 %b, %a"
    "  ret i32 %add"
    "}";

def intrinsic [llvm] binary + 20 (a:double, b:double):double
"define double @${mangledName}(double %a, double %b) nounwind uwtable readnone {"
"  %add = fadd double %a, %b"
"  ret double %add"
"}";

def intrinsic [llvm,always_inline] binary = 2 (a:ref(double), b:double):ref(double)
"define internal ${rettype} @${mangledName}(${argtype0} %${argname0}, ${argtype1} %${argname1}) nounwind uwtable { "
"  store ${argtype1} %${argname1}, ${argtype0} %${argname0} "
"  ret ${rettype} %${argname0} "
"}";

# Define ':' for sequencing: as a low-precedence operator that ignores operands
# and just returns the RHS.
def binary : 1 (x:double, y:double) y;
def binary : 1 (x:double, y:void) y;
def binary : 1 (x:void, y:double) y;
def binary : 1 (x:double, y:ref(double)) y;
def binary : 1 (x:ref(double), y:double) y;

def test(x:double)
  printdd(x) :
  x = x + 1.0 :
  printdd(x);

test(123.0);
