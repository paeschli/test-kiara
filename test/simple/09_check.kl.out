value = 123.000000
value = 124.000000
